### PR TITLE
Adding Graphical User Interface to Game

### PR DESCRIPTION
**Pull request by ironsigma**
Added a separate Main class to handle the GUI.

It was difficult to use the existing view/controller interfaces since they are more geared towards a more procedural model.

So instead I created an event driven game board (BoardPanel) that send out messages when it needs to know the status of a square, and notifies when a piece is moved. This decouples the GUI from the engine, and neither need to know how they're implemented.

Overall very basic GUI, but enough to start building on. Let me know if you have any question or comments.

### CODE REVIEW

File: src/main/java/com/imsavva/checkers/server/gui/GraphicInterface.java [hunk 1]
```diff
@@ -1,22 +1,115 @@
...
-public class GraphicInterface extends JFrame {
     public GraphicInterface() {
+        initGame();
+        initMenuBar();
         initUI();
+        game.startGame();
+    }
+
+    private void initGame() {
+        board = new Board();
```
Comment (ironsigma) [inline src/main/java/com/imsavva/checkers/server/gui/GraphicInterface.java:28]: This is where the engine is initialized

File: src/main/java/com/imsavva/checkers/server/gui/GraphicInterface.java [hunk 2]
```diff
@@ -1,22 +1,115 @@
...
-    public void initUI() {
-        add(new BoardPanel().initComponents());
+    private void initMenuBar() {
+        JMenuBar menuBar = new JMenuBar();
+
+        JMenu fileMenu = new JMenu("File");
+        fileMenu.setMnemonic(KeyEvent.VK_F);
+        menuBar.add(fileMenu);
+
+        JMenuItem newGameMenuItem = new JMenuItem("New Game", KeyEvent.VK_N);
```
Comment (ironsigma) [inline src/main/java/com/imsavva/checkers/server/gui/GraphicInterface.java:43]: New game menu item doesn't do anything right now, but added it as an example.

File: src/main/java/com/imsavva/checkers/server/gui/GraphicInterface.java [hunk 3]
```diff
@@ -1,22 +1,115 @@
...
         setResizable(false);
 
         setTitle("Corners Game");
         setLocationRelativeTo(null);
-        setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
+        setDefaultCloseOperation(WindowConstants.EXIT_ON_CLOSE);
+    }
+
+    public BoardPanelPiece getPieceAt(int x, int y) {
+        Cell cell = board.getCellAt(x, y);
```
Comment (ironsigma) [inline src/main/java/com/imsavva/checkers/server/gui/GraphicInterface.java:72]: This is where the board asks the engine for the pieces in the squares.

File: src/main/java/com/imsavva/checkers/server/gui/GraphicInterface.java [hunk 4]
```diff
@@ -1,22 +1,115 @@
...
+        return BoardPanelPiece.builder()
+                .square(new Point(x, y))
+                .isEmpty(cell.isEmpty())
+                .isWhite(cell.isEmpty() || cell.getFigure().getColor() == Figure.Color.WHITE)
+                .build();
+    }
+
+    public void movingPiece(String from, String to) {
+        try {
+            game.move(from, to);
```
Comment (ironsigma) [inline src/main/java/com/imsavva/checkers/server/gui/GraphicInterface.java:82]: This is where the board notifies that a piece has been moved, we check for end game state here too.

File: src/main/java/com/imsavva/checkers/server/gui/GraphicInterface.java [hunk 5]
```diff
@@ -1,22 +1,115 @@
...
+        }
+    }
+
+    @Override
+    public void actionPerformed(ActionEvent actionEvent) {
+        if (actionEvent.getActionCommand().equals("Quit")) {
+            dispatchEvent(new WindowEvent(this, WindowEvent.WINDOW_CLOSING));
+
+        } else if (actionEvent.getActionCommand().equals("New Game")) {
+            log.info("Starting new game...");
```
Comment (ironsigma) [inline src/main/java/com/imsavva/checkers/server/gui/GraphicInterface.java:111]: This is where we handle menu events, to be implemented.

**Comment by savva-k** [comment]
@ironsigma, I love it, thank you for your contribution!

**Pull request merged**